### DISCUSSION TITLE
Structure Size Scaling

### DISCUSSION THREAD

**Discussion by JOgbebor** [CederGroupHub/smol#482] [Q&A]
Hello,

I am using smol to compute cluster concentrations from a Potts subspace. For my specific application, I would ideally like to analyze structures containing upwards of 5,000 atoms. I realize that the code is primarily intended to process DFT structures, and that my use case is out of the ordinary. Naturally, I've run into a bottleneck with regards to the speed of computation. To quantify the scaling, I performed two tests for the `subspace.corr_from_structure()` function (plots below). In both cases, the Potts subspace contained four orbits (one-, two-, three-, and four-atom clusters) loaded from a pre-saved subspace. The number of threads was set using `subspace.num_threads = N_threads` and timing was measured using `time.perf_counter()`. My main question is: what can I do smol-wise to make processing large structures more efficient, apart from modifying the number of threads? Thanks in advance.

Runtime vs system size (8 threads):
N [atoms] | Runtime [s] 
32, 0.04037
108, 0.24085
256,  0.18334
500, 13.55407
864, 273.13657
1372, 173.36762
2048, 5827.09123
2916, 13465.58817
![Screenshot 2024-07-31 182002](https://github.com/user-attachments/assets/2a592cf4-c83c-4e2a-a150-4b93d04d21b4)

Runtime vs number of threads (1372 atoms):
![Screenshot 2024-07-31 182220](https://github.com/user-attachments/assets/50faec4d-4121-4989-8dc3-92742cf2fd39)

**Comment by kamronald** [comment]
Hi @JOgbebor thanks for running this benchmark, the large jump in computation time at N > 1500 is quite concerning. I think the best thing you can do right now is to process these structures in parallel - there is an example in the jupyter notebook: smol/docs/src/notebooks/adding-structures-in-parallel.ipynb
Don't think there's another way to speed this up sadly, with the current implementation.
